proparse/src/main/java/org/prorefactor/proparse/antlr4/PostLexer.java

```diff
@@ -0,0 +1,305 @@
+/*******************************************************************************
+ * Copyright (c) 2003-2015 John Green
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    John Green - initial API and implementation and/or initial documentation
+ *******************************************************************************/ 
+package org.prorefactor.proparse.antlr4;
+
+import antlr.TokenStreamException;
+import antlr.RecognitionException;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.IntStream;
+import org.antlr.v4.runtime.ListTokenSource;
+import org.antlr.v4.runtime.TokenFactory;
+import org.antlr.v4.runtime.TokenSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.io.IOException;
+
+/**
+ * This class deals with &amp;IF conditions by acting as a filter between the lexer and the parser
+ */
+public class PostLexer implements TokenSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostLexer.class);
+
+  private final Lexer lexer;
+  private final ProgressLexer prepro;
+
+  private final LinkedList<PreproIfState> preproIfVec = new LinkedList<>();
+  private ProToken currToken;
+
+  public PostLexer(Lexer lexer) {
+    this.lexer = lexer;
+    this.prepro = lexer.getPreprocessor();
+  }
+
+  @Override
+  public ProToken nextToken() {
+    LOGGER.trace("Entering nextToken()");
+    try {
+      for (;;) {
+
+        getNextToken();
+
+        switch (currToken.getType()) {
+
+          case PreprocessorParser.AMPIF:
+            preproIf();
+            break; // loop again
+
+          case PreprocessorParser.AMPTHEN:
+            // &then are consumed by preproIf()
+            throwMessage("Unexpected &THEN");
+            break;
+
+          case PreprocessorParser.AMPELSEIF:
+            preproElseif();
+            break; // loop again
+
+          case PreprocessorParser.AMPELSE:
+            preproElse();
+            break; // loop again
+
+          case PreprocessorParser.AMPENDIF:
+            preproEndif();
+            break; // loop again
+
+          default:
+            return currToken;
+
+        }
+      }
+    } catch (IOException | RecognitionException | TokenStreamException caught) {
+      throw new RuntimeException(caught);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS00112)